Feature/setup associations

Zafron047 opened the pull request:
### Dear Reviewer,

Please review the code changes and relationships implemented to meet the project requirements.

### Pull Request: Implementing Classroom-Student and Person-Book Relationships

**Description:**

This pull request addresses the project requirements by implementing the following key features:

1. **Classroom and Student Relationship:**
   - Created a `Classroom` class with an `@label` instance variable and getter/setter methods.
   - Established a "has-many" relationship between `Classroom` and `Student`, ensuring that a classroom can have multiple students and that each student belongs to a classroom.

2. **Book and Person many-tomany Relationship Rental class:**
   - Created a `Book` class with `@title` and `@author` instance variables and getter/setter methods.
   - Created a `Rental` class with an `@date` instance variable and getter/setter methods.
   - Implemented a "has-many" relationship between `Book` and `Rental`, allowing a book to have multiple rentals.
   - Implemented a "belongs-to" relationship between `Rental` and `Book`, ensuring that each rental belongs to a book.
   - Implemented a "has-many" relationship between `Person` and `Rental`, allowing a person to have multiple rentals.
   - Implemented a "belongs-to" relationship between `Rental` and `Person`, ensuring that each rental belongs to a person.
   - Modified the constructor of `Rental` to accept `Book` and `Person` objects, setting these associations during initialization.

**Pull Request Changes:**
- Added new classes `Classroom`, `Student`, `Book`, `Rental`, and `Person`.
- Implemented relationships between classes as described in the project requirements.
- Modified the `Rental` constructor to accept `Book` and `Person` objects for better association handling.

Thank you for your review and feedback!

Regards,

@Zafron047

Zafron047 committed:
Create Classroom class

Zafron047 committed:
Create has-many side in Classroom class

Zafron047 committed:
Create has-many/belongs-to relationship between Classroom and Student

Zafron047 committed:
Create Book, Rental classes

Zafron047 committed:
Create has-many relation between Book and Rental*

Zafron047 committed:
Create relations: 1) Person has-many Rentals 2) Rental belongs-to Person

Zafron047 committed:
Fix linter errors

Zafron047 committed:
Fix add_rental methods in book and rental classes

Zafron047 committed:
Add test cases and fix age instance as integer in Person class

rloterh commented on book.rb:
```diff
@@ -0,0 +1,14 @@
+class Book
+  attr_accessor :title, :author
+  attr_reader :rentals
```
- [x] `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ it. Kindly ensure to improve upon this aspect.

Zafron047 replied:
```ruby
class Book
  attr_accessor :title, :author, :rentals

``` 
Resolved issue.

Zafron047 committed:
Fix linter errors

Zafron047 commented:
Dear Reviewer,

I forgot to add the test cases in `test.rb` file which I prepared earlier. I have updated the `test.rb` file for your convenience. So that you can run a thorough testing of the class outputs. And I have also fixed linter errors.

Although I did these changes after submitting the pull request. So I sincerely beg your pardon if any inconvenience due to this occurred during your check.

Sincerely,

@Zafron047

rloterh commented on person.rb:
```diff
@@ -1,17 +1,18 @@
 require_relative 'nameable'
 
 class Person < Nameable
+  attr_accessor :name, :age
+  attr_reader :id, :rentals
```
- [x] Here again `rental` should be marked with `attr_accessor`, as we intend to both _access (get)_ as well as _write (set)_ it. Kindly ensure to improve upon this aspect.

Zafron047 replied:
```ruby
class Person < Nameable
  attr_accessor :name, :age, :rentals

``` 
Resolved issue.

rloterh reviewed (changes requested):
Hi Zafron, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there!
🏁 🚦 ⛪ 🚜 🏃 🌴 🏫 🌳 🏘️

## Highlights 👍  
- Good job so far with creating associations among objects!
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

Cheers and Happy coding! 👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Zafron047 committed:
Change rentals instance to both getter and setter in Book and Person classes

Zafron047 commented:
Dear @rloterh 
appreciate you for reviewing my code and raising issue. I have resolved the code blocks.

```ruby
class Book
  attr_accessor :title, :author, :rentals

``` 
```ruby
class Person < Nameable
  attr_accessor :name, :age, :rentals

``` 

Thanks 😸

bienvenuushindi reviewed (approved):
## Status: Approved :heavy_check_mark: :heavy_check_mark:

Hi @Zafron047     

Your project is complete! Great job for making all the required changes 🥇. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

## Good Points 👍
- The repo looks professional.:heavy_check_mark:
- There are no linter errors:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:
- Your project looks good:heavy_check_mark:

<br>
Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

Zafron047 commented:
@bienvenuushindi Thanks a lot for reviewing and approving the pr. 😄
